### PR TITLE
[SPARK-37841][SQL] BasicWriteTaskStatsTracker should not try get status for a skipped file

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In the comment https://github.com/apache/spark/pull/35117#issuecomment-1007171965 in #35117 and other details, we noticed that the BasicWriteTaskStatsTracker keeps complaining that some of the written files are missing. This is truely worrisome.

```scala
scala> spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_1")
22/01/07 14:22:48 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.
``` 

This happens for ORC format and the root cause is because, in OrcOutputWriter, the `recordWriter` and the `file` are lazily generated once have a record to write. So this step is skipped when the records iterator is empty. In the `commit` part, we will use the `file` to update stats assuming it exists. This causes `FileNotFoundException` and results not stats update. Then, `numSubmittedFiles` and `numFiles` become inconsistent and issue a false warning like above.

In this PR, we make the hive.OrcOutputWriter and OrcFileFormat behave the same as other data sources. That said, it will always create a file even when there are no records at all. 

### Why are the changes needed?

fix a regression
- the warning is wrong
- make builtin data sources behave same

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

added ut

locally

```log
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.0-SNAPSHOT
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 11.0.12)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.range(3).write.option("compression", "zstd").orc("./spark-warehouse/zstdtest_23")

scala> spark.range(3).write.option("compression", "zstd").orc("./spark-warehouse/zstdtest_24")

scala> 
```

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala [hunk 1]
```diff
@@ -142,8 +142,13 @@ class BasicWriteTaskStatsTracker(
     numSubmittedFiles += 1
   }
 
-  override def closeFile(filePath: String): Unit = {
-    updateFileStats(filePath)
+  override def closeFile(filePath: String, isPathCreated: Boolean): Unit = {
+    if (isPathCreated) {
+      updateFileStats(filePath)
+    } else {
+      logDebug(s"$filePath is not pre-touched by writer, skipping update file stats")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:149]:
```suggestion
      logDebug(s"$filePath is not created due to no data, skip updating file stats")
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 2]
```diff
@@ -305,6 +305,12 @@ private[orc] class OrcOutputWriter(
       recordWriter.close(Reporter.NULL)
     }
   }
+
+  /**
+   * If `recordWriterInstantiated` is false, the output file is not pretouched.
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:310]: Alternatively, can we fix the Hive ORC data source to always write the file? This seems wrong to me. We should at least write one file even if the input query is empty, to record the output schema.

Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:310]: You are right, we shall make this writer consistent with all the others

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 3]
```diff
@@ -299,11 +292,26 @@ private[orc] class OrcOutputWriter(
...
+      case NonFatal(e) => log.warn(e.toString, e)
+    }
+    recordWriter.close(Reporter.NULL)
+  }
+
+  private def getOrCreateInternalWriter(): Writer = {
+    val writerField = recordWriter.getClass.getDeclaredField("writer")
+    writerField.setAccessible(true)
+    var writer = writerField.get(recordWriter).asInstanceOf[Writer]
+    if (writer == null) {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:307]: do we need to consider multi-thread here?

Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala:307]: org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat.OrcRecordWriter init `writer` while writing the first element without synchronizing and we call `write` and `close` serially. I guess it's safe here

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Review by HyukjinKwon — COMMENTED** [review]
This is actually a very old issue SPARK-15474 and SPARK-8501. The change LGTM

**Pull request closed without merging**